### PR TITLE
Absyn line tracking

### PR DESCRIPTION
**Pull request by guardbotmk3**
Adds support for line, column, offset tracking in the Abstract Syntax classes.
Because the location support in cup 0.11a is spotty, 0.11a is almost 12 years old, and location support didn't work until recently, I decided that for location support, we would require cup 0.11b-2014-06-11 or later. I verified that without the -l option, cup 0.11a can still process the resulting files.
2552ae9 contains some edits besides the reversion that allow it to work with 0.11a when line number support is off.

We'll have to work something out with travis to test the -l option only with a newish cup.

### CODE REVIEW

**Comment by guardbotmk3** [comment]
Ping?

File: source/src/BNFC/Backend/Java.hs [hunk 1]
```diff
@@ -352,7 +353,7 @@ cf2AntlrLex' l = CF2Lex
 
 -- | CF -> PARSER GENERATION TOOL BRIDGE
 -- | function translating the CF to an appropriate parser generation tool.
-type CF2ParserFunction = String -> String -> CF -> SymEnv -> String
+type CF2ParserFunction = String -> String -> CF -> Bool -> SymEnv -> String
```
Comment (andreasabel) [inline source/src/BNFC/Backend/Java.hs]: Please don't use plain `Bool`. Define a 2-value data type with telling names instead.  Or at least define a type synonym.

File: source/src/BNFC/Backend/Java/CFtoAntlr4Parser.hs [hunk 2]
```diff
@@ -53,8 +53,9 @@ type MetaVar     = (String, Cat)
 
 -- | Creates the ANTLR parser grammar for this CF.
 --The environment comes from CFtoAntlr4Lexer
-cf2AntlrParse :: String -> String -> CF -> SymEnv -> String
-cf2AntlrParse packageBase packageAbsyn cf env = unlines
+--The bool is line numbering, which is currently not supported.
```
Comment (andreasabel) [inline source/src/BNFC/Backend/Java/CFtoAntlr4Parser.hs]: Put argument documentations down as haddock comments, i.e, a `-- ^` comment following the argument type.  See https://github.com/andreasabel/haskell-style-guide/blob/master/haskell-style.md

**Comment by andreasabel** [comment]
Thanks for the contribution!
One of my missions is to improve the code quality of bnfc, in terms of documentation, maintainability, automatic testing, etc.  Here is a style guide for Haskell code (WIP): https://github.com/andreasabel/haskell-style-guide/blob/master/haskell-style.md

**Comment by guardbotmk3** [comment]
OK, Bool has been excised from the places where it was used previously, and the C++ Backend has been updated as well. I can change the name of the line numbering flag if you'd like, but at least it's not a boolean now.

**Comment by andreasabel** [comment]
Thanks for the change.

Er, sorry to harp on this, but `LineNumber` sounds like you are passing a line number.  Could it be something like `HaveLineNumbers`, `HaveLocations`, `HaveLoc`, `HavePositions`, `HavePos` which indicates a flag?

**Comment by guardbotmk3** [comment]
OK, I named it RecordPositions. I even renamed all the 'ln' variables rp. Thanks grep and sed.

**Comment by guardbotmk3** [comment]
The Travis failure is the same one that I've seen before about GHC versions. The 2 non-broken versions passed.

**Comment by guardbotmk3** [comment]
I re-ran the tests and encountered some errors, that I've cleaned up with this push. Everything should be in good order.

**Comment by andreasabel** [comment]
Thanks, I merged this.
I resolved some conflicts with my own patches done for #221, which I pushed now.

**Pull request merged**